src/storage.py

```diff
@@ -0,0 +1,267 @@
+
+        # enrolls available disks with WAL/DB and save osd data.
+        with guard(self.charm, self.name):
+            self.charm.status.set(MaintenanceStatus("Enrolling OSDs"))
+            self._enroll_with_wal_db(disk=enroll["disk"], wal=enroll["wal"], db=enroll["db"])
+            self.charm.status.set(ActiveStatus())
+
+    def _on_osd_devices_attached(self, event: StorageAttachedEvent):
+        """Event handler for storage attach event."""
+        if not microceph._is_ready():
+            logger.warning("MicroCeph not ready yet, deferring storage event.")
+            event.defer()
+            return
+
+        self._clean_stale_osd_data()
+
+        enroll = []
+        for storage in [device for device in self.juju_storage_list() if "osd-devices" in device]:
+            path = self.juju_storage_get(storage_id=storage, attribute="location")
+            if not self._get_osd_num(path, "osd-devices"):
+                enroll.append(path)
+
+        with guard(self.charm, self.name):
+            self.charm.status.set(MaintenanceStatus("Enrolling OSDs"))
+            self._enroll_disks_in_batch(enroll)
+            self.charm.status.set(ActiveStatus())
+
+    def _on_storage_detaching(self, event: StorageDetachingEvent):
+        """Updates the attached storage devices in state."""
+        # check if the wal/db device is being used.
+        osd_num = self._get_osd_num(event.storage.location.as_posix(), event.storage.name)
+
+        # If detaching wal/db disk is used with an OSD.
+        if osd_num:
+            with guard(self.charm, self.name):
+                try:
+                    self.remove_osd(osd_num)
+                except CalledProcessError as e:
+                    if self._is_safety_failure(e.stderr):
+                        warning = f"Storage {event.storage.full_id} detached, provide replacement for osd.{osd_num}."
+                        logger.warning(warning)
+                        # clean records since juju will deprovision device.
+                        self.remove_osd(osd_num, force=True)
+                        raise BlockedExceptionError(warning)
+
+    """helpers"""
+
+    def _is_safety_failure(self, err: str) -> bool:
+        """Checks if the subprocess error is caused by safety check."""
+        return "need at least 3 OSDs" in err
+
+    def _run(self, cmd: list) -> str:
+        """Wrapper around subprocess run for storage commands."""
+        process = run(cmd, capture_output=True, text=True, check=True, timeout=180)
+        logger.debug(f"Command {' '.join(cmd)} finished; Output: {process.stdout}")
+        return process.stdout
+
+    def _enroll_with_wal_db(self, disk: list, wal: list, db: list):
+        """Checks if sufficient devices are available to be enrolled into OSDs."""
+        enrollment_count = min(len(disk), len(wal), len(db))
+        for i in range(enrollment_count):
+            try:
+                microceph.add_osd_cmd(disk[i], wal[i], db[i])
```

Comment:
what we could do in an event driven way is to define a `storage_updated` event on the StorageHandler, which the charm can subscribe to. The charm can perform client notification in the 'storage_updated' handler.

Reply:
wdyt @lmlg @sabaini ?

Reply:
I wonder could we not just perform client notifications on the spot? IMHO if we just have this use case (informing clients via notification) I'd call YAGNI on adding an event. If it later turns out we want to react to this event in several places we can always refactor

Reply:
My concern is around the responsibilities of the Storage class. Currently `StorageHandler` only contains juju storage handling logic. This way the storage event handlers don't have worry about stuff like clients and who needs to be notified of what. Adding client notification here would (only my opinion) add unneeded complexity.

Reply:
With a custom event, the storage handler would only need to `.emit()` an event with relevant information (current state of OSDs) whenever a storage change happens. This logic is relatively simpler for storage context and leaves space for dedicated client notification logic in the charm (which `StorageHandler` need not worry about)

Reply:
@sabaini how convincing was I ? (I've used up all of my rizz)

Reply:
:-D 

The reasoning for YAGNI is usually to focus on the current requirements to avoid unnecessary complexity, and rely on refactoring to evolve a design if and when needed. That is, do the simplest thing that could possibly work, and rely on refactoring (and testing) to improve the design when we find we need it. It's a matter of balance of course, this shouldn't mean to discourage forward looking design. Maybe it's enough to centralize all concerns of updating clients in a separate function or class? But these are just my 0.2c, I'm certainly not going to block over this

Reply:
> But these are just my 0.2c
Don't say that man I value you guys's opinion too much. 😭

Reply:
I propose a mid-point, since this notification thingy needs to be hooked to `add-disk` action too. How about we do this as a separate task ? That will give us space to accommodate both juju-storage and `add-disk` based OSDs and how (or what) needs to be notified to the clients ?